src/ds/logger.h

```diff
@@ -134,13 +171,72 @@ namespace logger
   {
     bool operator==(LogLine& line)
     {
-      std::cout << line.ss.str() << std::flush;
+      write(line.ss.str());
 
       if (line.log_level == Level::FATAL)
         throw std::logic_error("Fatal: " + line.ss.str());
 
       return true;
     }
+
+    static void write(const std::string& s)
+    {
+      // When logging from host code, print local time.
+      ::timespec ts;
+      ::timespec_get(&ts, TIME_UTC);
+      std::tm now;
+      ::localtime_r(&ts.tv_sec, &now);
+
+      std::cout << fmt::format(
+                     "{:%Y-%m-%d %H:%M:%S}.{:0<6}        ",
+                     now,
+                     ts.tv_nsec / 1000)
+                << s << std::flush;
+    }
+
+    static void write(const std::string& s, size_t ms_offset_from_start)
+    {
+      // When logging messages received from the enclave, print local time,
+      // and the offset to time inside the enclave at the time the message
+      // was logged there.
+      // Not thread-safe (uses std::localtime)
+      ::timespec ts;
+      ::timespec_get(&ts, TIME_UTC);
+      std::tm now;
+      ::localtime_r(&ts.tv_sec, &now);
+      time_t elapsed_s = ms_offset_from_start / 1000;
+      ssize_t elapsed_ns = (ms_offset_from_start % 1000) * 1000000;
+
+      // Enclave time is recomputed every time. If multiple threads
+      // log inside the enclave, offsets may not always increase
+      ::timespec enclave_ts{logger::config::start.tv_sec + elapsed_s,
+                            logger::config::start.tv_nsec + elapsed_ns};
+      if (enclave_ts.tv_nsec > 1000000000)
+      {
+        enclave_ts.tv_sec++;
+        enclave_ts.tv_nsec -= 1000000000;
+      }
+
+      // We assume time in the enclave is delayed compared to time on the
```

Comment:
I think `delayed` is unclear here, we think the enclave time is _earlier_ than the host time, right?

Reply:
Well time is delayed in the sense that the internal clock is effectively running late, when the enclave thinks it's 10.50.45.300, it's probably already 10.50.45.302. So the enclave time value is always smaller than the host value. I will rephrase that because it's clearly unclear!